Edit as a subtask midpoint updates

adamwhitlock1 opened the pull request:
## Summary

- Adds a breadcrumb component to the edit page and edit fallback page
  - custom styling using styled components to try and minimize bloat for this single component 
- Adds a modal to confirm that a user wants to cancel their changes if they have unsaved edits to a field
- New prop for `ProfileInformationFieldController`, that allows a custom cancel modal component to be provided
  - Lets us use this 'god' component for this edit page and hook up the custom modal with specific actions that can be taken when the user takes action on the modal buttons.
- `EditContext` created to allow the `EditConfirmCancelModal` to access the onCancel function as it is an indirect relationship of the edit page.
- styling edits based on midpoint review
  - links for alert when contact info is missing are now using the va-link with active property set to make them bolded and styled as required with chevron hover animation
  - added 'NOTIFICATION SETTINGS' text above field heading
  - cleaned up some minor styling around padding and margin of the page, mostly for adding the breadcrumb and other text
- My team, the Authenticated Experience team owns maintenance of this component and related work
- This all lives behind the 'profile_use_field_editing_page` feature toggle and is only live in staging at this point.

Update: Added mock endpoint for POST/PUT requests to update the email-address. The requests will now work correctly when using the edit page. It is worth noting that after adding an email when one was previously missing, the page will redirect to the notification settings but you wont see that change reflected, since the data doesn't persist in the mock server.

## Related issue(s)

- https://github.com/department-of-veterans-affairs/va.gov-team/issues/59987
- https://github.com/department-of-veterans-affairs/va.gov-team/issues/62077 (minor styling edits)

## Testing done

- Tested locally with feature toggles in various states.
- Added/updated unit tests to Edit, EditBreadcrumb, and NotificationSettings
- Added E2E test to missing-contact-infor-mobile test
- Added e2e test for the cancel modal on edit page
- You can view the modal by running vets website locally and toggling on the profile_use_field_editing_page and then hitting the url: http://localhost:3001/profile/edit?fieldName=mobilePhone&returnPath=%2Fprofile%2Fnotifications
- Also need to run the mock server
- Make changes to the field, and then click the cancel button.
    - Modal should then appear and allow you to return to editing, or cancel editing and will be redirected to the notification preferences page.
- to test the fallback page content you can visit the root edit url with no params: http://localhost:3001/profile/edit

## Screenshots

|         | Before | After |
| ------- | ------ | ----- |
| Edit page | ![Screenshot 2023-08-01 at 12 56 14 PM](https://github.com/department-of-veterans-affairs/vets-website/assets/8332986/14fd8a0f-67f2-4225-9210-42a55b4c1044) | ![Screenshot 2023-08-01 at 12 51 17 PM](https://github.com/department-of-veterans-affairs/vets-website/assets/8332986/c5f93203-5ab9-4c63-8f39-23053c4a0686) |
| Edit page fallback | ![Screenshot 2023-08-01 at 12 55 41 PM](https://github.com/department-of-veterans-affairs/vets-website/assets/8332986/6ab931fd-b29e-49df-bd57-203385d9ba69) | ![Screenshot 2023-08-01 at 12 52 54 PM](https://github.com/department-of-veterans-affairs/vets-website/assets/8332986/2ae4163c-1259-4849-9fd8-5b25c45fe37d) |
| missing contact info alert | ![Screenshot 2023-08-01 at 1 02 37 PM](https://github.com/department-of-veterans-affairs/vets-website/assets/8332986/bbb65dfc-f872-496e-a5dc-79e8980187b9) | ![Screenshot 2023-08-01 at 12 52 33 PM](https://github.com/department-of-veterans-affairs/vets-website/assets/8332986/469fdf75-676d-4dde-befb-3910817f5cbd) |
| cancel unsaved changes modal | NONE | ![Screenshot 2023-08-01 at 12 51 31 PM](https://github.com/department-of-veterans-affairs/vets-website/assets/8332986/db0ec3b8-fbd6-49c5-aaf8-0937a1341495) |

## What areas of the site does it impact?

Profile > Notification Settings

## Acceptance criteria

- [x] Modal added
- [x] Alert content and link styling updated
- [x] Breadcrumb added
- [x] Other midpoint review updates made

### Quality Assurance & Testing

- [x] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [x] Linting warnings have been addressed
- [x] Screenshot of the developed feature is added

### Error Handling

- [x] Browser console contains no warnings or errors.

Commit:
feat: add cancel confirm modal, context for edit, breadcrumb

Commit:
feat: extract breadcrumb, update content

Commit:
test: add unit for breadcrumb, cleanup

Commit:
rename unit test

Commit:
fix: clear state of edit modal on cancel

Commit:
chore: comment cleanup

allisonlu commented:
## general comments/questions:

* When I spin up the mock server w `profileUseFieldEditingPage` set to true and `profileShowEmailNotificationSettings` set to false, and user with no email nor phone in contact info, I only see option to Add phone number to profile, but am I expecting to see options to add both phone and email? (I'm comparing it to the [Sketch file](https://www.sketch.com/s/afd69a1f-72d2-430b-9b62-285e9d3f479c/p/8F33D983-75FB-4317-9DAF-D8306F3966FB/canvas?annotation=c65ad1c1-750b-41f4-bbd8-cc92c7d0da8e))
* I see expected behavior when I both feature toggles mentioned above are set to true
<img width="532" alt="Screenshot 2023-08-01 at 2 07 29 PM" src="https://github.com/department-of-veterans-affairs/vets-website/assets/8542413/cad929c7-5f70-479e-a862-db7687d7f81c">

---

* While trying to update email and save, I get this error in terminal:
> `put /v0/profile/email_addresses : "mock not initialized for token: _global put /v0/profile/email_addresses"` 

* Also getting warning in (browser console):
> Failed prop type: Invalid prop `error` of type `object` supplied to `ConfirmRemoveModal`, expected `string`.

* tested with users: `loa3UserWithNoMobilePhone`, `loa3UserWithNoEmail`, `loa3UserWithNoEmailOrMobilePhone`, `loa3User72`
<img width="541" alt="Screenshot 2023-08-01 at 2 15 20 PM" src="https://github.com/department-of-veterans-affairs/vets-website/assets/8542413/f9b6b440-5c11-460c-94f8-4bd57886182f">

---

* Getting warning in browser console: Failed prop type: The prop `disabledForCheckbox` is marked as required in `NotificationChannel`, but its value is `undefined`.

adamwhitlock1 commented:
Thanks @allisonlu for taking a look at this. I have mocked out the POST for the mobile phone save, but not the email POST. I will add that to the mock server.

Strange that you only saw the one link in the alert for phone and not email with that.
There are a lot of toggles / data cases that are at play, and I should do a better job of figuring out the explicit ones to use to simulate each case and document that.

If `profileUseFieldEditingPage: true` and `profileShowEmailNotificationSettings: true`, as well as the user mock is set to be `loa3UserWithNoEmailOrMobilePhone` then that should show the alert with both links, but maybe I have messed something up along the way.

Sounds like I should make an e2e test for this that covers the alert with one link and two links to cover our bases. I'll get one added to this PR.

Maybe tomorrow we can get on a quick call and you could help me debug the issue?

Commit:
chore: cleanup mock server

Commit:
test: add e2e for edit page

Commit:
feat: add mock POST and PUT for email address updates

Commit:
Merge branch 'main' into Profile/aw/59987/edit-subtask-midpoint-updates

adamwhitlock1 commented:
I've got some breaking tests that I'm looking into on this PR. Will update those tests and make sure this branch is up to date with main.

Commit:
fix: revert export structure of user mocks slightly

Commit:
Merge branch 'main' into Profile/aw/59987/edit-subtask-midpoint-updates

Commit:
test: fix e2e tests

Commit:
fix: update more user objects in tests

Commit:
Merge branch 'main' into Profile/aw/59987/edit-subtask-midpoint-updates

Commit:
Merge branch 'main' into Profile/aw/59987/edit-subtask-midpoint-updates

Commit:
Merge branch 'main' into Profile/aw/59987/edit-subtask-midpoint-updates

Commit:
Merge branch 'main' into Profile/aw/59987/edit-subtask-midpoint-updates

adamwhitlock1 commented:
@allisonlu and @dellerbie This is finally 're-ready' for review. I updated some things to make it easier to run the mock server with the user endpoint and feature toggle values purely editing in the server.js file.

Right now with the configuration of the mock server. The alert for both missing contact types should show on the notifications page by default, and by modifying the user response you should be able to see the other cases.

I also added some e2e tests for that and the cancel confirmation modal to cover more of that, which of course pushed the LOC up, but hopefully its not too much to take a look over.

Commit:
fix: correct prop types

allisonlu reviewed (approved):
changes look good @adamwhitlock1 !! :shipit:

adamwhitlock1 commented on src/applications/personalization/profile/tests/components/notification-settings/NotificationSettings.unit.spec.jsx:
```diff
@@ -133,11 +133,9 @@ describe('<NotificationSettings />', () => {
 
-    expect(view.queryByText('Add an email address to your profile')).to.not
-      .exist;
+    expect(view.queryByTestId('add-email-address-link')).to.not.exist;
```
Good question! I like to use the table here to help me https://testing-library.com/docs/react-testing-library/cheatsheet/

`queryBy` doesn't throw an error when it doesn't find something vs all the rest throw errors, so that has been my go to for when I am checking that something does not exist in the UI.
